File: sdks/go/providers/murf/http.go
package murf

import (
	"bytes"
	"context"
	"errors"
	out "github.com/speechswitch/client/sdks/go/generated/murf_output"
	"github.com/speechswitch/client/sdks/go/runtime"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
)

type httpStream struct {
	parent, ctx                                                          context.Context
	cancel                                                               context.CancelFunc
	stop                                                                 func() bool
	transport                                                            runtime.HTTPTransport
	response                                                             *runtime.HTTPResponse
	body                                                                 runtime.Input[[]byte]
	limit                                                                int
	gen2, timed, inline                                                  bool
	once                                                                 sync.Once
	mutex, lifecycle                                                     sync.Mutex
	closed, initialized, received, audioDone, timingDone, done, terminal bool
	closeError                                                           error
	generation                                                           generation
}

func (s *httpStream) closeResources() error {
	s.once.Do(func() {
		s.cancel()
		s.lifecycle.Lock()
		s.closed = true
		body := s.body
		s.body = nil
		s.lifecycle.Unlock()
		if body != nil {
			s.closeError = body.Close()
		}
	})
	return s.closeError
}
func (s *httpStream) Close() error { s.stop(); return s.closeResources() }
func (s *httpStream) replaceBody(body runtime.Input[[]byte]) error {
	s.lifecycle.Lock()
	old, closed := s.body, s.closed
	if !closed {
		s.body = body
	}
	s.lifecycle.Unlock()
	if old != nil {
		_ = old.Close()
	}
	if closed && body != nil {
		_ = body.Close()
		return s.ctx.Err()
	}
	return nil
}
func readDocument(ctx context.Context, body runtime.Input[[]byte], limit int) ([]byte, error) {
	var data []byte
	for {
		chunk, err := body.Next(ctx)
		if err == io.EOF {
			return bytes.TrimPrefix(data, []byte{239, 187, 191}), nil
		}
		if err != nil {
			return nil, err
		}
		if len(chunk) > limit-len(data) {
			return nil, errors.New("Murf response exceeds MaxJSONBytes")
		}
		data = append(data, chunk...)
	}
}
func responseError(ctx context.Context, response *runtime.HTTPResponse, limit int) error {
	if response.StatusCode >= 200 && response.StatusCode < 300 {
		return nil
	}
	data, err := readDocument(ctx, response.Body, limit)
	if err != nil {
		return err
	}
	failure := &Error{StatusCode: runtime.Some(response.StatusCode), Body: runtime.DecodeUTF8(data)}
	if values, present := response.Header["Retry-After"]; present {
		failure.RetryAfter = runtime.Some(strings.Join(values, ", "))
	}
	return failure
}
func (s *httpStream) Next(ctx context.Context) (out.SynthesisItem, error) {
	stop := context.AfterFunc(ctx, func() { s.closeResources() })
	defer stop()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal || s.done {
		return nil, io.EOF
	}
	item, err := s.next(ctx)
	if cause := ctx.Err(); cause != nil {
		item, err = nil, cause
	} else if cause := s.parent.Err(); cause != nil {
		item, err = nil, cause
	} else if s.ctx.Err() != nil && !s.done {
		item, err = nil, io.EOF
	}
	if err != nil {
		s.terminal = true
		s.Close()
	}
	return item, err
}
func (s *httpStream) next(ctx context.Context) (out.SynthesisItem, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if s.ctx.Err() != nil {
		return nil, io.EOF
	}
	if !s.initialized {
		s.initialized = true
		if err := responseError(ctx, s.response, s.limit); err != nil {
			return nil, err
		}
		if s.gen2 {
			data, err := readDocument(ctx, s.response.Body, s.limit)
			if err != nil {
				return nil, err
			}
			s.generation, err = decodeGeneration(data, s.timed, s.inline)
			if err != nil {
				return nil, err
			}
			if err = s.replaceBody(nil); err != nil {
				return nil, err
			}
			if !s.inline {
				target, err := endpoint(s.generation.url, "", false)
				if err != nil {
					return nil, err
				}
				u, _ := url.Parse(target)
				if u.Scheme != "https" {
					return nil, errors.New("Murf returned an unsafe audio file URL")
				}
				request, err := http.NewRequestWithContext(s.ctx, http.MethodGet, target, nil)
				if err != nil {
					return nil, err
				}
				// The audio file is an independent HTTPS resource; never copy API credentials.
				s.response, err = runtime.OpenResponse(request, s.transport)
				if err != nil {
					return nil, err
				}
				if err = s.replaceBody(s.response.Body); err != nil {
					return nil, err
				}
				if err = responseError(ctx, s.response, s.limit); err != nil {
					return nil, err
				}
			}
		} else {
			contentType := strings.ToLower(strings.TrimSpace(strings.SplitN(s.response.Header.Get("Content-Type"), ";", 2)[0]))
			if contentType != "" && !strings.HasPrefix(contentType, "audio/") && contentType != "application/octet-stream" {
				return nil, errors.New("Murf returned a non-audio streaming response")
			}
		}
	}
	if !s.audioDone {
		var data []byte
		var err error
		if s.gen2 && s.inline {
			data, s.audioDone = s.generation.audio, true
			s.generation.audio = nil
		} else {
			data, err = s.response.Body.Next(ctx)
			if err == io.EOF {
				s.audioDone = true
				if err = s.replaceBody(nil); err != nil {
					return nil, err
				}
			} else if err != nil {
				return nil, err
			}
		}
		if len(data) > 0 {
			s.received = true
			if s.timed {
				return out.SynthesisItemAsOrderedOrTimeline{Value: out.MurfEnvelope{Correlation: out.MurfEnvelopeCorrelationAsTimeline{}, Audio: runtime.Some(data), Timestamps: []out.MurfTimestamp{}}}, nil
			}
			return out.SynthesisItemAsBytes{Value: data}, nil
		}
	}
	if !s.received {
		return nil, errors.New("Murf returned no audio")
	}
	if s.timed && !s.timingDone {
		s.timingDone = true
		return out.SynthesisItemAsOrderedOrTimeline{Value: out.MurfEnvelope{Correlation: out.MurfEnvelopeCorrelationAsTimeline{}, DurationMs: runtime.Some(s.generation.duration), Timestamps: s.generation.timestamps}}, nil
	}
	done := out.DoneEvent{}
	if s.gen2 {
		done.RemainingCharacters, done.Warning = runtime.Some(s.generation.remaining), s.generation.warning
	}
	s.done = true
	s.Close()
	return out.SynthesisItemAsDone{Value: done}, nil
}

File: sdks/rust/src/providers/murf/http.rs
use super::{
    protocol::{self, failure, Audio, Error},
    url, SynthesisItem,
};
use crate::{
    generated::murf_output::*,
    http::{HttpRequest, HttpResponse, HttpTransport, TransportError},
    runtime::StreamingInput,
};
use std::{
    future::{poll_fn, Future},
    pin::Pin,
    task::{Context, Poll},
};

struct Completion {
    duration: f64,
    timestamps: Vec<MurfTimestamp>,
    done: DoneEvent,
}
struct Prepared {
    body: Option<StreamingInput<Vec<u8>>>,
    inline: Option<Vec<u8>>,
    completion: Option<Completion>,
}
type Preparation<'a> = Pin<Box<dyn Future<Output = Result<Prepared, TransportError>> + Send + 'a>>;
pub(super) struct HttpStream<'a> {
    prepare: Option<Preparation<'a>>,
    ready: Option<Prepared>,
    timed: bool,
    received: bool,
    timing_done: bool,
    pub done: bool,
}
impl<'a> HttpStream<'a> {
    pub fn new(
        response: HttpResponse,
        transport: &'a dyn HttpTransport,
        gen2: bool,
        timed: bool,
        inline: bool,
        limit: usize,
    ) -> Self {
        Self {
            prepare: Some(Box::pin(prepare(
                response, transport, gen2, timed, inline, limit,
            ))),
            ready: None,
            timed,
            received: false,
            timing_done: false,
            done: false,
        }
    }
    pub fn next(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Option<SynthesisItem>, TransportError>> {
        if let Some(future) = &mut self.prepare {
            let ready = match future.as_mut().poll(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(v) => v?,
            };
            self.prepare = None;
            self.ready = Some(ready);
        }
        let ready = self.ready.as_mut().unwrap();
        let data = if let Some(bytes) = ready.inline.take() {
            Some(bytes)
        } else if let Some(body) = &mut ready.body {
            match body.as_mut().poll_next(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Some(Err(err))) => return Poll::Ready(Err(err)),
                Poll::Ready(Some(Ok(bytes))) if bytes.is_empty() => {
                    cx.waker().wake_by_ref();
                    return Poll::Pending;
                }
                Poll::Ready(Some(Ok(bytes))) => Some(bytes),
                Poll::Ready(None) => {
                    ready.body = None;
                    None
                }
            }
        } else {
            None
        };
        if let Some(bytes) = data.filter(|v| !v.is_empty()) {
            self.received = true;
            let item = if self.timed {
                let mut e = protocol::envelope(true);
                e.audio = Some(bytes);
                SynthesisItem::OrderedOrTimeline(e)
            } else {
                SynthesisItem::Bytes(bytes)
            };
            return Poll::Ready(Ok(Some(item)));
        }
        if !self.received {
            return Poll::Ready(Err(failure("Murf returned no audio")));
        }
        if self.timed && !self.timing_done {
            self.timing_done = true;
            let completion = ready.completion.as_mut().unwrap();
            let mut e = protocol::envelope(true);
            e.duration_ms = Some(completion.duration);
            e.timestamps = std::mem::take(&mut completion.timestamps);
            return Poll::Ready(Ok(Some(SynthesisItem::OrderedOrTimeline(e))));
        }
        let done = ready.completion.take().map_or(
            DoneEvent {
                event: Default::default(),
                remaining_characters: None,
                warning: None,
            },
            |c| c.done,
        );
        self.done = true;
        Poll::Ready(Ok(Some(SynthesisItem::Done(done))))
    }
}
async fn document(
    body: &mut StreamingInput<Vec<u8>>,
    limit: usize,
) -> Result<String, TransportError> {
    let mut data = Vec::new();
    poll_fn(|cx| {
        for _ in 0..128 {
            match body.as_mut().poll_next(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
                Poll::Ready(Some(Ok(bytes))) => {
                    if bytes.len() > limit - data.len() {
                        return Poll::Ready(Err(failure("Murf response exceeds max_json_bytes")));
                    }
                    data.extend(bytes);
                }
                Poll::Ready(None) => {
                    return Poll::Ready(Ok(String::from_utf8_lossy(
                        data.strip_prefix(&[239, 187, 191]).unwrap_or(&data),
                    )
                    .into_owned()))
                }
            }
        }
        cx.waker().wake_by_ref();
        Poll::Pending
    })
    .await
}
async fn check(response: &mut HttpResponse, limit: usize) -> Result<(), TransportError> {
    if !(200..300).contains(&response.status) {
        return Err(Box::new(Error {
            status: Some(response.status),
            body: document(&mut response.body, limit).await?,
            retry_after: response
                .headers
                .iter()
                .find(|(k, _)| k.eq_ignore_ascii_case("retry-after"))
                .map(|(_, v)| v.clone()),
        }));
    }
    Ok(())
}
async fn prepare(
    mut response: HttpResponse,
    transport: &dyn HttpTransport,
    gen2: bool,
    timed: bool,
    inline: bool,
    limit: usize,
) -> Result<Prepared, TransportError> {
    check(&mut response, limit).await?;
    if !gen2 {
        let content_type = response
            .headers
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case("content-type"))
            .map_or("", |(_, v)| v.as_str())
            .split(';')
            .next()
            .unwrap_or("")
            .trim()
            .to_ascii_lowercase();
        if !content_type.is_empty()
            && !content_type.starts_with("audio/")
            && content_type != "application/octet-stream"
        {
            return Err(failure("Murf returned a non-audio streaming response"));
        }
        return Ok(Prepared {
            body: Some(response.body),
            inline: None,
            completion: None,
        });
    }
    let text = document(&mut response.body, limit).await?;
    drop(response);
    let g = protocol::generation(&text, timed, inline)?;
    let completion = Some(Completion {
        duration: g.duration,
        timestamps: g.timestamps,
        done: DoneEvent {
            event: Default::default(),
            remaining_characters: Some(g.remaining),
            warning: g.warning,
        },
    });
    match g.audio {
        Audio::Bytes(bytes) => Ok(Prepared {
            body: None,
            inline: Some(bytes),
            completion,
        }),
        Audio::Url(target) => {
            let target = url(&target, "", false)?;
            if !target
                .split_once("://")
                .unwrap()
                .0
                .eq_ignore_ascii_case("https")
            {
                return Err(failure("Murf returned an unsafe audio file URL"));
            }
            // Audio files are independent resources; never forward API credentials.
            let mut response = transport
                .send(HttpRequest {
                    method: "GET".into(),
                    url: target,
                    headers: vec![],
                    body: vec![],
                })
                .await?;
            check(&mut response, limit).await?;
            Ok(Prepared {
                body: Some(response.body),
                inline: None,
                completion,
            })
        }
    }
}
